Support irreducible control flow by converting it to reducible control flow.

cfallin opened the pull request:
Wasm supports expressing only reducible control flow via its structured control flow mechanisms (forward-branches out of blocks and backward-branches to loop headers, in addition to structured if/else diamonds).

waffle accepts an arbitrary CFG as its IR, and this CFG is perfectly capable of expressing irreducible CFGs. For example, the canonical irreducible CFG

```plain
      1
    /   \
   /     \
  2 ----> 3
    <---
```

(that is, the graph with edges 1->2, 1->3, 2->3, 3->2) cannot be expressed in Wasm as direct control flow. Instead, a compiler that targets Wasm needs to either use indirection -- `i32` values as target tags that are fed into switches (`br_table`s), usually -- or code duplication. For the former case, we could build a loop body out of blocks 2 and 3 that accepts a header parameter that selects between 2 or 3; this "lowered loop" iterates twice as many times as the original loop. For the latter case, we could duplicate either block 2 or 3 and then "fall into" the main loop, for example by doing:

```plain
         1
    _  /   \
   | \/     \
   | 2 <---- 3'
   | |
   | |
   | 3
   \_/

```

The advantage of the label-variable approach is that it does not duplicate code; it adds a small constant overhead to code size per irreducible loop instead. The disadvantage is that it adds runtime cost, which is a potentially significant performance cliff if an unexpected irreducible side-edge is added for some reason. In addition, it has some significant implementation complexity:

- All entrances to a loop are indirected through the one header block. When we support blockparams and perform the reducibility transform on the CFG, we need to find the *union* of all blockparam signatures of all side-entrances and rewrite block targets to use "default" zero-values for the gaps (unused parameters for the dynamically-selected target). This is very awkward, and potentially impossible when we have non-nullable values (e.g. GC refs) in the future.

- When a side-entrance to a *nested* loop occurs, we need to indirect through multiple header blocks. This adds complexity to the transform too because we need to "fixpoint" the transform somehow: the branch targets of the header block's `br_table` also need to be transformed.

In this PR, instead, I opted for the code-duplication approach. It has at most 2x code-size overhead (duplicate all but the header block in a loop, if we have a side-entrance one block later in the loop body), but has zero runtime overhead. In addition, it avoids the complexities above.

This PR adds the Reducifier, a transform whose approach is described in a module-level comment in `backend/reducify.rs`. In brief, the idea is to:

1. Compute an approximate loop-nest over the CFG, computing "header sets" for every block that describe the loops the blocks belong to. We do this by first computing a reverse postorder, then finding RPO backedges (edges from rpo-index `i` to rpo-index `j` such that `j <= i`), adding the header (target with lower RPO index) to all header-sets "under" the edge (rpo-indices from `i` to `j` inclusive), then fixing up loop nesting in one final pass over the header-sets in RPO order with a stack that tracks the nest, "pushing out" the extent of outer loops to cover inner loops that start within their extent.

   In the above canonical irreducible CFG, blocks 2 and 3 will have header-sets of either `{2}` or `{3}`, depending on which of the two valid reverse postorders the initial RPO computation chose (which depends on which out-edges are visited first).

3. Find irreducible edges given these header sets: these are edges from `B1` to `B2` such that header `H` is in `headers(B2)` but not `headers(B1)`, and `H != B2`. (Intuitively, a "side-entrance".) If no such edges exist, the CFG is reducible; early-out without modifications.

4. Perform a context-sensitive code-duplication pass. We define a *context set* per block that is like the header set, but the dual: it denotes "side-entered loops". (Formally, moving across an edge `(From, To)`, we compute `ToContext = headers(To) & !{To} & (FromContext U !headers(From))`: this adds to the context when we side-enter a loop, and removes from the context when we jump back to the loop header or leave the loop.)

   As part of this, we perform a limited "max-SSA" pass on a set of cut-blocks computed as side-entrances and exits (that is, exactly blocks where we change context), so that when we duplicate code, we don't have to reason about fixing up SSA across stapled-together duplicated code fragments.

This results in a reducible CFG, which we can then compile with the rest of waffle's backend.

cfallin committed:
Support irreducible control flow by converting it to reducible control flow.

Wasm supports expressing only reducible control flow via its structured
control flow mechanisms (forward-branches out of blocks and
backward-branches to loop headers, in addition to structured if/else
diamonds).

waffle accepts an arbitrary CFG as its IR, and this CFG is perfectly
capable of expressing irreducible CFGs. For example, the canonical
irreducible CFG

```plain
      1
    /   \
   /     \
  2 ----> 3
    <---
```

(that is, the graph with edges 1->2, 1->3, 2->3, 3->2) cannot be
expressed in Wasm as direct control flow. Instead, a compiler that
targets Wasm needs to either use indirection -- `i32` values as target
tags that are fed into switches (`br_table`s), usually -- or code
duplication. For the former case, we could build a loop body out of
blocks 2 and 3 that accepts a header parameter that selects between 2 or
3; this "lowered loop" iterates twice as many times as the original
loop. For the latter case, we could duplicate either block 2 or 3 and
then "fall into" the main loop, for example by doing:

```plain
         1
    _  /   \
   | \/     \
   | 2 <---- 3'
   | |
   | |
   | 3
   \_/

```

The advantage of the label-variable approach is that it does not
duplicate code; it adds a small constant overhead to code size per
irreducible loop instead. The disadvantage is that it adds runtime cost,
which is a potentially significant performance cliff if an unexpected
irreducible side-edge is added for some reason. In addition, it has some
significant implementation complexity:

- All entrances to a loop are indirected through the one header block.
  When we support blockparams and perform the reducibility transform on
  the CFG, we need to find the *union* of all blockparam signatures of
  all side-entrances and rewrite block targets to use "default"
  zero-values for the gaps (unused parameters for the
  dynamically-selected target). This is very awkward, and potentially
  impossible when we have non-nullable values (e.g. GC refs) in the
  future.

- When a side-entrance to a *nested* loop occurs, we need to indirect
  through multiple header blocks. This adds complexity to the transform
  too because we need to "fixpoint" the transform somehow: the branch
  targets of the header block's `br_table` also need to be transformed.

In this PR, instead, I opted for the code-duplication approach. It has
at most 2x code-size overhead (duplicate all but the header block in a
loop, if we have a side-entrance one block later in the loop body), but
has zero runtime overhead. In addition, it avoids the complexities
above.

This PR adds the Reducifier, a transform whose approach is described in
a module-level comment in `backend/reducify.rs`. In brief, the idea is
to:

1. Compute an approximate loop-nest over the CFG, computing "header
   sets" for every block that describe the loops the blocks belong to.
   We do this by first computing a reverse postorder, then finding RPO
   backedges (edges from rpo-index `i` to rpo-index `j` such that `j <=
   i`), adding the header (target with lower RPO index) to all
   header-sets "under" the edge (rpo-indices from `i` to `j` inclusive),
   then fixing up loop nesting in one final pass over the header-sets in
   RPO order with a stack that tracks the nest, "pushing out" the extent
   of outer loops to cover inner loops that start within their extent.

   In the above canonical irreducible CFG, blocks 2 and 3 will have
   header-sets of either `{2}` or `{3}`, depending on which of the two
   valid reverse postorders the initial RPO computation chose (which
   depends on which out-edges are visited first).

2. Find irreducible edges given these header sets: these are edges from
   `B1` to `B2` such that header `H` is in `headers(B2)` but not
   `headers(B1)`, and `H != B2`. (Intuitively, a "side-entrance".) If no
   such edges exist, the CFG is reducible; early-out without
   modifications.

3. Perform a context-sensitive code-duplication pass. We define a
   *context set* per block that is like the header set, but the dual: it
   denotes "side-entered loops". (Formally, moving across an edge
   `(From, To)`, we compute `ToContext = headers(To) & !{To} &
   (FromContext U !headers(From))`: this adds to the context when we
   side-enter a loop, and removes from the context when we jump back to
   the loop header or leave the loop.)

   As part of this, we perform a limited "max-SSA" pass on a set of
   cut-blocks computed as side-entrances and exits (that is, exactly
   blocks where we change context), so that when we duplicate code, we
   don't have to reason about fixing up SSA across stapled-together
   duplicated code fragments.

This results in a reducible CFG, which we can then compile with the rest
of waffle's backend.

cfallin commented:
@sunfishcode, would you be willing to review this?

Re: testing: I have a manually-written test case in the PR, but I intend to also write a fuzzing target that drives this with arbitrary CFGs and ensures that the Wasm backend can compile them.

cfallin committed:
Add fuzzing target.

cfallin commented:
Updated with a fuzz target (`irreducible`) and some fixes; gets through a few million random CFGs now with successful Wasm codegen.

sunfishcode commented on fuzz/fuzz_targets/irreducible.rs:
```diff
@@ -0,0 +1,95 @@
+        let mut edges_by_origin: PerEntity<Block, Vec<Block>> = PerEntity::default();
+        for &(from, to) in &self.edges {
+            let from = Block::from(u32::from(from) % num_blocks);
+            let to = Block::from(u32::from(to) % num_blocks);
```
Instead of `% num_blocks` for these two, would it make sense to make this [reject inputs](https://docs.rs/libfuzzer-sys/latest/libfuzzer_sys/macro.fuzz_target.html#rejecting-inputs) with out-of-range indices? That way the corpus won't end up with redundant CFGs.

cfallin replied:
Done!

cfallin committed:
Add warning about code-size blowup in the worst case to doc-comment.

Per @sunfishcode's offline comment -- the tradeoff makes sense for us
for now but we should explicitly document / warn about it.

sunfishcode commented on fuzz/fuzz_targets/irreducible.rs:
```diff
@@ -0,0 +1,95 @@
+#[derive(Clone, Debug, Arbitrary)]
+struct CFG {
+    num_blocks: u8,
+    edges: Vec<(u8, u8)>,
```
What kind of lengths does the `Arbitrary` implementation for `Vec` produce? If it's not long enough, we may not be fuzzing the full range of interesting possibilities.

cfallin replied:
I instrumented a run just now to see and we seem to be seeing edge-lists up to ~50 edges or more; and anecdotally the fuzzer was able to find some interesting cases with nested loops yesterday, so I think we're within the realm of good coverage at least!

sunfishcode reviewed (commented):
Looks good!

cfallin committed:
Review feedback on fuzzer.
